### ISSUE TITLE
[1837] Implementation List

### ISSUE THREAD

**Issue by crericha** [tobymao/18xx#11101]
General
- [x] Bank size: 14268
- [x] Currency format: `%dK`
- [x] Starting capital: `{ 3: 730, 4: 555, 5: 450, 6: 380, 7: 330 }`
- [x] Player count: 3 - 7
- [x] Train roster: Not in rulebook
- [x] Tile roster: Not in rulebook

Phases
- [x] 2
  - [x] 1 OR
  - [x] Yellow
  - [x] Train limit: 2/4/-
  - [x] Mountain pass hexes can only be played on by company/corporation whose director owns the corresponding mountain railway
  - [x] Cannot play tiles in Bosnia-Herzegovina
- [x] 3
  - [x] 2 ORs
  - [x] Yellow, Green
  - [x] 2Gs available (alongside 1Gs)
  - [x] Train limit: 2/3/-
  - [x] Mountain pass hexes cost 110 if not director of corresponding mountain railway
  - [x] Can cross-buy trains
- [x] 3+1
  - [x]  Same as 3, but train limit 1/3/-
- [x] 4
  - [x] 2 ORs
  - [x] Yellow, Green
  - [x] 3Gs available. Unsold 1G and 2G trains still available
  - [x] Train limit 1/3/4
- [x] 5
  - [x] 3 ORs
  - [x] Yellow, Green, Brown
  - [x] 4Gs available. Unsold 1G, 2G, 3G trains still available.
  - [x] Train limit -/2/3

Events:
- [x] 4
  - [x] Southern railway formed
  - [x] KK may start (KK 1 director decides)
  - [x] Italy no longer playable. Tiles removed from Italy.
- [x] 4E 
  - [x] Hungarian may form (UG 1 director decides)
- [x] 4+1
  - [x] KK must form if not already
  - [x] KK minors must convert
- [x] 5
  - [x] Hungarian must form if not already
  - [x] Hungarian minors must convert
  - [x] Remaining coal companies are exchanged for shares
  - [x] Mountain Railways are closed

Stock Market
- [x] Hex stock market

Trains
- [x] 3G rusts 1G
- [x] 4G rusts 2G
- [x] 4 rusts 2
- [x] 5 rusts 3
- [x] 5+2 rusts 3+1
<img width="391" alt="image" src="https://github.com/user-attachments/assets/2476a990-55f8-4411-a023-3cead3ad5b4e">

Mountain Railways
- [x] Privates each with 110K discount on a specific hex
- [x] Count for cert limit
- [x] Some come with a minor
- [x] Close on purchase of first 5
<img width="392" alt="image" src="https://github.com/user-attachments/assets/50315df0-817b-4ff3-aec5-f9877cd95446">

Minor Companies
- [x] Can own goods and passenger trains
- [x] Not required to own a train
- [x] Only have home token
- [x] Southern minors
<img width="390" alt="image" src="https://github.com/user-attachments/assets/5d870ad6-c8d9-473d-ab5b-25b59d7a1e4a">

- [x] KK minors
<img width="393" alt="image" src="https://github.com/user-attachments/assets/0a90cfef-5576-4bb9-ad56-773cb1bd6cde">

- [x] Hungarian minors (UG)
<img width="396" alt="image" src="https://github.com/user-attachments/assets/34b59ce0-e446-4ef7-84b9-b8a2b75b988d">

- [x] UG 1 and 2 have 4 shares

Coal Companies
- [x] Start with 100
- [x] Can only own goods trains ("G")
- [x] Not required to own a train
- [x] Come with an option on share of a major
<img width="401" alt="image" src="https://github.com/user-attachments/assets/0a753987-8d2c-472d-9cd4-5ab2e4e4de04">

Major Corporations
- [x] Only two corps may use each par price
- [x] 50% to float (does not include exchange shares)
- [x] Treasury is num IPO shares + par price

National Railways
- [x] DIrector's certificate is 10% share
- [x] Starting values are pre-determined
Both
- [x] Must own a train, even if no route
<img width="395" alt="image" src="https://github.com/user-attachments/assets/e3d95d5d-2b4c-4c05-8f60-0e29e8f01b60">

Initial Auction
- [x] Start packet
<img width="849" alt="image" src="https://github.com/user-attachments/assets/64286659-f98e-4917-b5ae-47bbf804f633">

- [x] Selection auction
- [x] Minimum value on share
- [x] Increments of 5
- [x] Ends when all pass
- [x] Next player to act is the player after the purchaser

Stock Round
- [x] First stock round
  - [x] Unsold start packet items available at face 
  - [x] Ug secondary shares for minors 1 and 3 are now available
- [x] Available shares
  - [x] Once the start packet is fully sold, shares of major companies are available 
for purchase (latest edition says also the Ug secondary shares. Need to confirm)
  - [x] National shares also available, but will not operate until it forms
  - [x] Exchange shares reserved
- [x] sell-buy or buy-sell
- [x] 60% max ownership
- [x] Sell movement is diagonally down and left one (regardless of shares sold)
  - [x] Rules state two different movements. First share sold, diagonally down and left and subsequent shares sold down and left. I'm thinking the former is a typo.
  - [x] No longer moves if reaches bottom or left of chart
- [x] Shares of nationals can be 
- [x] Shares for corp in yellow portion of the stock market don't count against the cert limit
- [x] End of stock round movement
  - [x] sold out (no ipo or bank shares) diagonally up to the right if director owns 40% or less, otherwise diagonally up to the left

Operating Round
- [x] Exchanges before operation
  1. Coal companies may exchange
  2. KK and UG national railways may form or minors may be abosrbed
  3. Mountain railway owners are paid income
- [x] Order of operation
  1. Coal companies in this order - EPP, RGTE, EOD, EKT, MLB, ZKB, SPB, LRB, BB, EHS
  5. Southern minors (ascending order)
  6. KK minos (ascending order)
  7. Hungarian minors (ascending order)
  8. Major corporations (share price descending)
- [x] VLR (SD 5) must decide home city (Milan or Venice) before any track laid by any company. If not sold, defaults to Milan. (latest edition says its track lay, not any. Need to confirm correct ruling).
- [x] OR actions
  - [x] Lay or upgrade a tile
    - [x] Upgrades new track must be a continuation of existing path or upgrade city value (semi-restrictive)
    - [x] Bosnia-Herzegovina hexes cannot be built until phase 3
    - [x] Can lay in Italy in phases 2 and 3. At the start of phase 4 all tiles removed from Italy and cannot lay there for the rest of the game.
    - [x] Bozen and Bregenz become a lettered city in green. 
    - [x] The Bozen tile gets automatically placed at the start of phase 4.
  - [x] Place a token
    - [x] Home token placed when purchased/floated (different from original rules)
    - [x] Token costs are determine from home station.
    - [x] Can backtrack to lay stations (need to figure out exactly what that means)
    - [x] Majors: 20K per hex for 2nd token, and 40K per hex for 3rd.
    - [x] Nationals: 20K per hex for first, non-inherited token, 40K per hex for others
      - [x] Southern 5 minor token goes on charter, not hex (20K per hex)
      - [x] Southern home is Vienna
      - [x] KK tokens remain in Vienna until brown. Two returned once brown.
      - [x] KK home is Vienna
      - [x] Hungarian tokens remain in Budapest until upgrade to brown. Once upgraded, one token is returned.
      - [x] Hungrian home is Budapest.
  - [x] Run trains
    - [x] Vienna is considered one city, cannot run twice with same train
    - [x] Buda and Pest considered separate cities
    - [x] Passenger trains
      - [x] Cannot run coal mines
      - [x] Can start/end at a town
      - [x] E trains skip towns
    - [x] Goods trains
      - [x] Must start at the coal mine (doesn't count as a stop).
      - [x] Cannot include another coal mine
      - [x] Must include a station (coal mine is a coal company's station)
      - [x] Runs unlimited towns
      - [x] A route must contain more than just the coal mine
      - [x] The coal mine revenue goes to the treasury. The remainder is regular earnings.
      - [x] Coal mine value increases in brown
  - [x] Pay dividends - full, half, or hold
    - [x] Coal companies: always half pay
    - [x] Minor companies: half pay (rounded down of total earnings)
    - [x] Major and National: 
      - [x] full, half, hold. (rounded down for half)
      - [x] Bank shares pay the company
      - [x] Shares exchanged after operating do not get paid, instead goes to the bank (NEEDS TESTING)
  - [x] Adjust share price
      - [x] Share price moves right on full payout. (Up and the the right if can't move, following the arrow)
      - [x] Share price moves diagonally down and right if possible on half pay
      - [x] Share price moves left on hold or no zero revenue. If not possible, then diagonally down and left.
  - [x] Train purchase
    - [x] Trains can be voluntarily surrendered by paying half its purchase price to the bank.
    - [x] Surrendered trains go to the bank pool
  - [x] EMR
    - [x] May buy any available train
    - [x] Can only sell shares from operating corp if it doesn't cause a ownership change
    - [x] Can cross buy for any amount that doesn't bankrupt the director
  - [x] Bankruptcy
    - [x] Shares to the bank pool
    - [x] Bank finances the remainder for the train (needs to be repaid)
    - [x] Corp cannot pay dividends until bank repaid
    - [x] Cannot cross buy train until debt repaid
    - [x] Player with most shares of company takes ownership, priority order if tied
    - [x] Ownerless corporation can have additional trains financed the bank if trainless again

Coal Company Exchange
- [x] Option provided at beginning of SR/OR once at least 50% of the major bought
- [x] Must exchange if no shares remain in the major's IPO or upon purchase of first 5T
- [x] Exchange order is dealt with in operating order of associated majors
- [x] All assets go to the major
- [x] After all exchanges, majors must get under the train limit. Must discard newly received trains first. Cannot discard existing trains.
- [x] If coal company already operated
  - [x] Cannot operate trains this OR
  - [x] Exchanged shares do not pay this OR
- [x] Corp still receives all assets if it hasn't floated yet (first 5T purchased)

National Railway Formation
- [x] Optional formations can happen at beginning of SR or OR
- [x] Does not run that OR if formed via a phase change
- [x] Minor company certs exchanged for 10% of national
- [x] All minor assets go to national
- [x] Director tie goes to owner of lowest minor number
- [x] Southern:
  - [x] SD5 token returned to railway
  - [x] Receives 710K in treasury
  - [x] Starts at 142 stock value
- [x] KK:
  - [x] KK 1 decides if it forms at phase 4. If no, asks again start of each SR/OR. Must form at 4+1.
  - [x] If KK forms before 4+1, KK 2 and KK 3 can choose to fold in or continue to run. Asked again at start of each SR/OR and must fold in at 4+1. Asked in order from current director clockwise (confirm).
  - [x] Receives 840K in treasury
  - [x] Starts at 120 stock value
- [x] Hungarian:
  - [x] Ug 1 decides if it forms at phase 4E. If no, asks again start of each SR/OR. Must form at 5
  - [x] Ug 1 and 3 share owners must exchange when director of minors decides to fold in.
  - [x] If Ug forms before 5, Ug 2, etc. can choose to fold in or continue to run. Asked again at start of each SR/OR and must fold in at 5. Asked in order from current director clockwise (confirm).
  - [x] Receives 875K in treasury
  - [x] Starts at 175 stock value

Game End
- [x] Ends after current OR that the bank runs out
- [x] Game ends immediately if all but one player bankrupt

**Comment by ollybh** [comment]
>  Can backtrack to lay stations (need to figure out exactly what that means)

When working out where you can place tokens, you're allowed to go the wrong way at track junctions. This is most important for the BH major – it's fairly common for this company to be started by the player who owns the BB coal company. Before starting the BH the BB will have upgraded hex O19 to a #25 tile, with one path from the mine to the town in O17, and the other from the mine towards Sarajevo. On its first turn BH can place a tile in Sarajevo connecting to the #25 tile and then trace a route through O19, reversing at the junction and place a token it one of the cities closer to Vienna. This means it will be able to run the G-train inherited from BB for a decent amount on its first turn.

So you're probably going to have to have two graphs, one for tokens, and one for track and trains. And the token graph will have to be calculated differently, which could be a challenge.

**Comment by ollybh** [comment]
One point to the list for track laying: town upgrades are different from normal. Yellow single town tiles upgrade to a green tile with the existing path with a town, and a second path without a town. Yellow double town tiles upgrade to a green single town tile with a K/X/trident-arrangement that keeps the same exits that were on the yellow tile.

**Comment by ollybh** [comment]
For dividends from the four-share minors (Ug1 and Ug3), you round fractions when working out how much each player will earn, not how much each share will pay.

If Ug1 runs for 50K, then a player who owns one share (25%) will receive 12K, but if a player owns both shares (50%) they will receive 25K, not 24K.

**Comment by crericha** [comment]
> > Can backtrack to lay stations (need to figure out exactly what that means)
> 
> When working out where you can place tokens, you're allowed to go the wrong way at track junctions. This is most important for the BH major – it's fairly common for this company to be started by the player who owns the BB coal company. Before starting the BH the BB will have upgraded hex O19 to a #25 tile, with one path from the mine to the town in O17, and the other from the mine towards Sarajevo. On its first turn BH can place a tile in Sarajevo connecting to the #25 tile and then trace a route through O19, reversing at the junction and place a token it one of the cities closer to Vienna. This means it will be able to run the G-train inherited from BB for a decent amount on its first turn.
> 
> So you're probably going to have to have two graphs, one for tokens, and one for track and trains. And the token graph will have to be calculated differently, which could be a challenge.

@ollybh If the token graph thought the tiles were lawson, it'd maybe automatically do the backtracking?

**Comment by scottredracecar** [comment]
Without looking up the rule, I believe you can only do one "backtrack." I
have no ideas how to implement that, but treating as Lawson seems like a
fine thing to do in the meantime as a hack (ignoring the 1 back track
limit). The BH has a lot of trouble without the backtrack rule.

**Comment by crericha** [comment]
I happen to have the rules open. Looks like it is unlimited backtracking.

>  For the purpose of playing tokens only, the route between a corporation’s home station and the new one may involve backtracking (multiple times if necessary)

**Comment by ollybh** [comment]
> @ollybh If the token graph thought the tiles were lawson, it'd maybe automatically do the backtracking?

I don't know the internals of the graph walking code well, but I had assumed that it is the junction parts on Lawson-type tiles are what lets the graph walk head off in any direction, and these tiles will not have these.

For most tiles it would work if you assumed that there are invisible paths between all edges with track on a plain/town tiles (so on a #23 tile as well as the paths between edges 0 and 3 and 0 and 4, you have an extra one between edges 3 and 4). But that wouldn't work with tiles like #16 to #20 where there are two separate tracks without a junction to backtrack at.

**Comment by ianwilson156** [comment]
Re: backtracking: the original German rules were ambiguous, and Lonny gave contradictory answers when asked! We always play multiple backtracking, but you may want to make it an option (single vs multiple).

**Comment by ollybh** [comment]
@crericha I've just been poking around tiles defined by various game, and looking at 1837 I noticed that there's a mistake on the brown #&NoBreak;432 tile – there should be a path connecting the two cities on the tile, similar to the green Budapest tile (#&NoBreak;425). If you squint at this [image on BGG](https://boardgamegeek.com/image/128533/1837-rail-building-in-the-austro-hungarian-empire) you can see a very short piece of track between the cities.

It looks like you're making great progress on this! I'm looking forward to playing it.
